File: scratchpad/failed-jobs/2026-06-11__08-14-38/transformation-drop-invalid__HwbbzV2/artifacts/user/hookdeck-task/provision.js
const fs = require('fs');
const path = require('path');

const runId = process.env.ZEALT_RUN_ID;
const apiKey = process.env.HOOKDECK_API_KEY;

if (!runId || !apiKey) {
  console.error("Missing ZEALT_RUN_ID or HOOKDECK_API_KEY environment variables.");
  process.exit(1);
}

const apiBase = "https://api.hookdeck.com/2024-09-01";

async function cleanUp() {
  console.log("Starting clean-up of existing resources...");

  // 1. Clean Connections
  const connRes = await fetch(`${apiBase}/connections`, {
    headers: { 'Authorization': `Bearer ${apiKey}` }
  });
  if (connRes.ok) {
    const data = await connRes.json();
    const existingConns = data.models || [];
    for (const conn of existingConns) {
      if (conn.name === `conn-${runId}`) {
        console.log(`Deleting existing connection: ${conn.id}`);
        await fetch(`${apiBase}/connections/${conn.id}`, {
          method: 'DELETE',
          headers: { 'Authorization': `Bearer ${apiKey}` }
        });
      }
    }
  }

  // 2. Clean Sources
  const srcRes = await fetch(`${apiBase}/sources`, {
    headers: { 'Authorization': `Bearer ${apiKey}` }
  });
  if (srcRes.ok) {
    const data = await srcRes.json();
    const existingSrcs = data.models || [];
    for (const src of existingSrcs) {
      if (src.name === `src-${runId}`) {
        console.log(`Deleting existing source: ${src.id}`);
        await fetch(`${apiBase}/sources/${src.id}`, {
          method: 'DELETE',
          headers: { 'Authorization': `Bearer ${apiKey}` }
        });
      }
    }
  }

  // 3. Clean Destinations
  const destRes = await fetch(`${apiBase}/destinations`, {
    headers: { 'Authorization': `Bearer ${apiKey}` }
  });
  if (destRes.ok) {
    const data = await destRes.json();
    const existingDests = data.models || [];
    for (const dest of existingDests) {
      if (dest.name === `dst-${runId}`) {
        console.log(`Deleting existing destination: ${dest.id}`);
        await fetch(`${apiBase}/destinations/${dest.id}`, {
          method: 'DELETE',
          headers: { 'Authorization': `Bearer ${apiKey}` }
        });
      }
    }
  }

  // 4. Clean Transformations
  const transRes = await fetch(`${apiBase}/transformations`, {
    headers: { 'Authorization': `Bearer ${apiKey}` }
  });
  if (transRes.ok) {
    const data = await transRes.json();
    const existingTrans = data.models || [];
    for (const trans of existingTrans) {
      if (trans.name === `validate-${runId}`) {
        console.log(`Deleting existing transformation: ${trans.id}`);
        await fetch(`${apiBase}/transformations/${trans.id}`, {
          method: 'DELETE',
          headers: { 'Authorization': `Bearer ${apiKey}` }
        });
      }
    }
  }

  console.log("Clean-up completed.");
}

async function run() {
  try {
    console.log(`Starting Hookdeck provisioning for Run ID: ${runId}`);

    // Clean up first to ensure idempotency
    await cleanUp();

    // 1. Create Source
    const sourceName = `src-${runId}`;
    console.log(`Creating source: ${sourceName}...`);
    const sourceRes = await fetch(`${apiBase}/sources`, {
      method: 'POST',
      headers: {
        'Authorization': `Bearer ${apiKey}`,
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({
        name: sourceName
      })
    });

    if (!sourceRes.ok) {
      const errText = await sourceRes.text();
      throw new Error(`Failed to create source: ${sourceRes.status} ${errText}`);
    }

    const sourceData = await sourceRes.json();
    const sourceId = sourceData.id;
    console.log(`Created Source ID: ${sourceId}`);

    // 2. Create Destination
    const destName = `dst-${runId}`;
    console.log(`Creating destination: ${destName}...`);
    const destRes = await fetch(`${apiBase}/destinations`, {
      method: 'POST',
      headers: {
        'Authorization': `Bearer ${apiKey}`,
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({
        name: destName,
        url: "https://mock.hookdeck.com"
      })
    });

    if (!destRes.ok) {
      const errText = await destRes.text();
      throw new Error(`Failed to create destination: ${destRes.status} ${errText}`);
    }

    const destData = await destRes.json();
    const destId = destData.id;
    console.log(`Created Destination ID: ${destId}`);

    // 3. Create Transformation
    const transformName = `validate-${runId}`;
    const transformCode = `addHandler("transform", (request, context) => {
  const body = request.body;
  if (!body || typeof body !== "object") {
    console.log("validation_failed: user_id");
    return null;
  }
  if (typeof body.user_id !== "string" || body.user_id.trim() === "") {
    console.log("validation_failed: user_id");
    return null;
  }
  if (typeof body.amount !== "number" || !Number.isFinite(body.amount)) {
    console.log("validation_failed: amount");
    return null;
  }
  return request;
});`;

    console.log(`Creating transformation: ${transformName}...`);
    const transformRes = await fetch(`${apiBase}/transformations`, {
      method: 'POST',
      headers: {
        'Authorization': `Bearer ${apiKey}`,
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({
        name: transformName,
        code: transformCode
      })
    });

    if (!transformRes.ok) {
      const errText = await transformRes.text();
      throw new Error(`Failed to create transformation: ${transformRes.status} ${errText}`);
    }

    const transformData = await transformRes.json();
    const transformationId = transformData.id;
    console.log(`Created Transformation ID: ${transformationId}`);

    // 4. Create Connection
    const connName = `conn-${runId}`;
    console.log(`Creating connection: ${connName}...`);
    const connRes = await fetch(`${apiBase}/connections`, {
      method: 'POST',
      headers: {
        'Authorization': `Bearer ${apiKey}`,
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({
        name: connName,
        source_id: sourceId,
        destination_id: destId,
        rules: [
          {
            type: "transform",
            transformation_id: transformationId
          }
        ]
      })
    });

    if (!connRes.ok) {
      const errText = await connRes.text();
      throw new Error(`Failed to create connection: ${connRes.status} ${errText}`);
    }

    const connData = await connRes.json();
    console.log(`Created Connection ID: ${connData.id}`);

    // 5. Write output.log
    const logPath = "/home/user/hookdeck-task/output.log";
    const logContent = `Source ID: ${sourceId}\nTransformation ID: ${transformationId}\n`;
    fs.writeFileSync(logPath, logContent);
    console.log(`Successfully wrote log to ${logPath}`);

  } catch (error) {
    console.error("Error during provisioning:", error);
    process.exit(1);
  }
}

run();

File: jobs/2026-06-11__08-14-38/fanout-two-destinations-rate-lim__AgyNUEG/artifacts/user/hookdeck-fanout/provision.py
import os
import requests
import json
import sys

api_key = os.environ.get("HOOKDECK_API_KEY")
run_id = os.environ.get("ZEALT_RUN_ID")

if not api_key or not run_id:
    print("Error: HOOKDECK_API_KEY or ZEALT_RUN_ID not set")
    sys.exit(1)

headers = {
    "Authorization": f"Bearer {api_key}",
    "Content-Type": "application/json"
}
base_url = "https://api.hookdeck.com/2025-07-01"

source_name = f"fanout-src-{run_id}"
fast_dest_name = f"fanout-fast-{run_id}"
slow_dest_name = f"fanout-slow-{run_id}"
fast_conn_name = f"fanout-fast-conn-{run_id}"
slow_conn_name = f"fanout-slow-conn-{run_id}"

# 1. Clean up existing resources to avoid conflicts
print("Cleaning up existing connections...")
r = requests.get(f"{base_url}/connections", headers=headers)
if r.status_code == 200:
    for conn in r.json().get("models", []):
        if conn.get("name") in [fast_conn_name, slow_conn_name]:
            conn_id = conn.get("id")
            print(f"Deleting connection {conn.get('name')} ({conn_id})...")
            requests.delete(f"{base_url}/connections/{conn_id}", headers=headers)

print("Cleaning up existing destinations...")
r = requests.get(f"{base_url}/destinations", headers=headers)
if r.status_code == 200:
    for dest in r.json().get("models", []):
        if dest.get("name") in [fast_dest_name, slow_dest_name]:
            dest_id = dest.get("id")
            print(f"Deleting destination {dest.get('name')} ({dest_id})...")
            requests.delete(f"{base_url}/destinations/{dest_id}", headers=headers)

print("Cleaning up existing sources...")
r = requests.get(f"{base_url}/sources", headers=headers)
if r.status_code == 200:
    for src in r.json().get("models", []):
        if src.get("name") == source_name:
            src_id = src.get("id")
            print(f"Deleting source {src.get('name')} ({src_id})...")
            requests.delete(f"{base_url}/sources/{src_id}", headers=headers)


# 2. Create Source
print(f"\nCreating source {source_name}...")
src_payload = {
    "name": source_name,
    "type": "WEBHOOK"
}
r = requests.post(f"{base_url}/sources", headers=headers, json=src_payload)
print("Source creation status:", r.status_code)
if r.status_code not in [200, 201]:
    print("Error details:", r.text)
    sys.exit(1)
source_data = r.json()
source_id = source_data.get("id")
print(f"Created Source ID: {source_id}")


# 3. Create Fast Destination
print(f"\nCreating fast destination {fast_dest_name}...")
fast_dest_payload = {
    "name": fast_dest_name,
    "type": "MOCK_API",
    "config": {
        "rate_limit": None,
        "rate_limit_period": None
    }
}
r = requests.post(f"{base_url}/destinations", headers=headers, json=fast_dest_payload)
print("Fast Destination creation status:", r.status_code)
if r.status_code not in [200, 201]:
    print("Error details:", r.text)
    sys.exit(1)
fast_dest_data = r.json()
fast_dest_id = fast_dest_data.get("id")
print(f"Created Fast Destination ID: {fast_dest_id}")


# 4. Create Slow Destination
print(f"\nCreating slow destination {slow_dest_name}...")
slow_dest_payload = {
    "name": slow_dest_name,
    "type": "MOCK_API",
    "config": {
        "rate_limit": 2,
        "rate_limit_period": "second"
    }
}
r = requests.post(f"{base_url}/destinations", headers=headers, json=slow_dest_payload)
print("Slow Destination creation status:", r.status_code)
if r.status_code not in [200, 201]:
    print("Error details:", r.text)
    sys.exit(1)
slow_dest_data = r.json()
slow_dest_id = slow_dest_data.get("id")
print(f"Created Slow Destination ID: {slow_dest_id}")


# 5. Create Fast Connection
print(f"\nCreating fast connection {fast_conn_name}...")
fast_conn_payload = {
    "name": fast_conn_name,
    "source_id": source_id,
    "destination_id": fast_dest_id
}
r = requests.post(f"{base_url}/connections", headers=headers, json=fast_conn_payload)
print("Fast Connection creation status:", r.status_code)
if r.status_code not in [200, 201]:
    print("Error details:", r.text)
    sys.exit(1)
fast_conn_data = r.json()
fast_conn_id = fast_conn_data.get("id")
print(f"Created Fast Connection ID: {fast_conn_id}")


# 6. Create Slow Connection
print(f"\nCreating slow connection {slow_conn_name}...")
slow_conn_payload = {
    "name": slow_conn_name,
    "source_id": source_id,
    "destination_id": slow_dest_id
}
r = requests.post(f"{base_url}/connections", headers=headers, json=slow_conn_payload)
print("Slow Connection creation status:", r.status_code)
if r.status_code not in [200, 201]:
    print("Error details:", r.text)
    sys.exit(1)
slow_conn_data = r.json()
slow_conn_id = slow_conn_data.get("id")
print(f"Created Slow Connection ID: {slow_conn_id}")

print("\nAll resources successfully provisioned!")
print(f"source_id: {source_id}")
print(f"fast_destination_id: {fast_dest_id}")
print(f"slow_destination_id: {slow_dest_id}")
print(f"fast_connection_id: {fast_conn_id}")
print(f"slow_connection_id: {slow_conn_id}")

# Write to state file
state = {
    "source_id": source_id,
    "fast_destination_id": fast_dest_id,
    "slow_destination_id": slow_dest_id,
    "fast_connection_id": fast_conn_id,
    "slow_connection_id": slow_conn_id
}
with open("/home/user/hookdeck-fanout/state.json", "w") as f:
    json.dump(state, f, indent=2)
print("Wrote state to state.json")
